### DISCUSSION TITLE
Can't use v4 Get()/Set() Array behaviour in v5

### DISCUSSION THREAD

**Discussion by crodwell** [jackc/pgx#1542] [Q&A] [answered]
With  the AssignTo() and Set() funcs available v4 now gone, scanning works from a pgx Row and also with basic go native types such as string & int  but not native slices:
```
	var result pgtype.Array[string]
	pgxconn.QueryRow(context.Background(), "SELECT ARRAY['test1', 'test2']").Scan(&result)

	 // result:
	 // (pgtype.Array[string]) {
         //       Elements: ([]string) (len=2 cap=2) {
         //       (string) (len=5) "test1",
         //       (string) (len=5) "test2"
         //      },
         //      Dims: ([]pgtype.ArrayDimension) (len=1 cap=1) {
         //       (pgtype.ArrayDimension) {
         //        Length: (int32) 2,
         //        LowerBound: (int32) 1
         //       }
         //      },
         //      Valid: (bool) true
         //     }

	data := []string{"test1", "test2"}

	err := pgtype.NewMap().SQLScanner(data).Scan(&result)
         fmt.Println(err)
         //  cannot scan _text (OID 1009) in text format into []string

```

How can I do the equivalent of `result.Set(data)` as I used to be able to do in pgx v4?

**Comment by jackc** [comment]
What exactly are you trying to do? `AssignTo` and `Set` shouldn't be necessary anymore.

You shouldn't need `SQLScanner` unless you are using `database/sql` and if you are that's not a valid argument for a `sql.Scanner`.

Scanning query results directly to Go slices should work out of the box. For example from the test suite: 

https://github.com/jackc/pgx/blob/19039e6dd10a8e5d4d5f8cae683f9b4e7757609b/pgtype/array_codec_test.go#L326-L329

**Comment by crodwell** [comment]
Effectively in v4 I had a model for each table defined as a struct with each field a `pgtype`. The API would take a request struct bound from a JSON request - and transform that into an update to one or more models using `Set` methods.  I could then compare changes to model for auditing, and save model state.. A pretty common pattern in an MVC type framework.

Yes, I gathered the Scan is really for scanning database results into the pgtype and there isn't a way to set the pgtype from native go in a generic way anymore .

I'm currently headed down a path of creating wrappers like this which I don't think is ideal.

```// TextArray is a wrapper for pgtype.Array[string] allowing a .Set() method
type TextArray struct {
	pgtype.Array[string]
}

// Set makes up for a non-existent method in the pgx v5 lib that worke in v4
func (dst *TextArray) Set(value []string) {
	// construct dst manually
}
```

**Comment by crodwell** [comment, marked as answer]
I was successful using wrapper structs:
```package pgxtype

import (
	"fmt"
	"strconv"

	"github.com/jackc/pgx/v5/pgtype"
)

// Bool is a wrapper for pgtype.Bool allowing a .Set() method
type Bool struct {
	pgtype.Bool
}

// Set makes up for a non-existent method in the pgx v5 lib that worked in v4
func (dst *Bool) Set(src any) error {
```

The only negative is my code that used to get a Bool in the following way:
```
var myBool pgtype.Bool
value := myBool.Bool
```
now has to be referenced like so due to the conflicting naming of Bool and it's underlying value:
```
var myBool customTypes.Bool
value := customTypes.Bool.Bool
```

It would be really nice if `Set()` and `Get()` were reinstated in pgx/pgtype v5

**Comment by jackc** [comment]
It's unlikely that `Set` and `Get` will return. They were mostly a consequence of how the type system in v4 combined the concepts of a value container and a value encoder / decoder. That design had deficiencies which were resolved by v5 making `Codec` is an explicit concept.

However, v5 usually makes it easier to create application or domain types that can be directly used with pgx by exposing per PostgreSQL type interfaces. For your bool example, your type could implement `BoolScanner` and `BoolValuer` directly. This would remove the need to wrap pgtype.Bool altogether.
